File: backend/app/services/product_service.py
import uuid
from typing import Optional
from app.models import Category

def get_category_by_name(conn, name: str) -> Optional[Category]:
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM categories WHERE name = ?", (name,))
    row = cursor.fetchone()
    if row:
        return Category(**dict(row))
    return None

def create_or_update_category(conn, name: str, price: int, initial_entries: int = 0) -> Category:
    cursor = conn.cursor()
    
    existing = get_category_by_name(conn, name)
    
    if existing:
        cursor.execute(
            "UPDATE categories SET price = ?, total_entries = total_entries + ? WHERE id = ?",
            (price, initial_entries, existing.id)
        )
        conn.commit()
        return get_category_by_name(conn, name)
    else:
        cat_id = str(uuid.uuid4())
        cursor.execute(
            "INSERT INTO categories (id, name, price, total_entries, total_exits) VALUES (?, ?, ?, ?, ?)",
            (cat_id, name, price, initial_entries, 0)
        )
        conn.commit()
        return Category(id=cat_id, name=name, price=price, total_entries=initial_entries, total_exits=0)

def list_categories(conn) -> list[Category]:
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM categories ORDER BY name ASC")
    return [Category(**dict(row)) for row in cursor.fetchall()]

File: backend-node/src/services/productService.js
const { v4: uuidv4 } = require('uuid');

function getCategoryByName(db, name) {
  return db.prepare('SELECT * FROM categories WHERE name = ?').get(name) || null;
}

function createOrUpdateCategory(db, name, price, initialEntries = 0) {
  const existing = getCategoryByName(db, name);

  if (existing) {
    if (price > 0) {
      db.prepare(
        'UPDATE categories SET price = ?, total_entries = total_entries + ? WHERE id = ?'
      ).run(price, initialEntries, existing.id);
    } else {
      db.prepare(
        'UPDATE categories SET total_entries = total_entries + ? WHERE id = ?'
      ).run(initialEntries, existing.id);
    }
    return getCategoryByName(db, name);
  }

  const catId = uuidv4();
  db.prepare(
    'INSERT INTO categories (id, name, price, total_entries, total_exits) VALUES (?, ?, ?, ?, ?)'
  ).run(catId, name, price, initialEntries, 0);
  return getCategoryByName(db, name);
}

function listCategories(db) {
  return db.prepare('SELECT * FROM categories ORDER BY name ASC').all();
}

module.exports = { getCategoryByName, createOrUpdateCategory, listCategories };
